### ISSUE TITLE
Nginx fails to start due to missing ssl cert keys

### ISSUE THREAD

**Issue by fz9** [engintron/engintron#561]
Every time I restart the server, Nginx fails to start because of an error like this:

`2017/06/17 05:46:13 [emerg] 17868#17868: SSL_CTX_use_PrivateKey_file("/var/cpanel/ssl/installed/keys/beba5_aaeb3_77d0c54668e08eb71f1a33b7xxxxxx.key") failed (SSL: error:02001002:system library:fopen:No such file or directory:fopen('/var/cpanel/ssl/installed/keys/beba5_aaeb3_77d0c54668e08eb71f1a33b7xxxxxx.key','r') error:20074002:BIO routines:FILE_CTRL:system lib error:140B0002:SSL routines:SSL_CTX_use_PrivateKey_file:system lib)`

The way I have to fix this is do a search for that key file and then do a cp of that from customer account into /var/cpanel/ssl/installed/keys/

This keeps happening with different accounts and is causing unnecessary downtime after reboot and requires manual intervention. Why is this happening? This is a shared server, so there is literally no way of controlling how customers choose to add/delete their certificates. How can I overcome this problem?

CloudLinux  7.3 x86_64 xen hvm, cPanel, EA4.

**Comment by imagina** [comment]
I had the same problem a couple of times already with latest version. Very annoying. There is no option to regenerate https vhosts or something to fix it quickly.

**Comment by Friends4U** [comment]
Why would you want to restart your server?!?

**Comment by Mike981874** [comment]
Same issue, I just disabled Engintron for the time being since it causes nginx to not start, and with a cron setup to restart nightly (due to new certificates needing it?) it was a problem. This originally happened around a month ago and was very rare but now it's occurring every 2-5 days when the service restarts happen.

@Friends4U I'm assuming he means the services. I'm not sure of any way to get around new SSL certs not working without a restart to nginx.

**Comment by imagina** [comment]
Ok, in my case i found the problem.

It was ngintron cron requiring to read files outside of /home and we were limiting that option in php.ini with open_basedir

Remove open_basedir restriction in your main php.ini file and it should run fine.

**Comment by fevangelou** [comment]
Also make sure /etc/crontab's syntax is not broken. Just open the file via the Engintron WHM app and just hit save.

**Comment by jhawkins002** [comment]
We have the same error - I had begun to think we were the only ones! After coming up with a work-around we closed our issue report thread, so it's definitely great to hear the feedback from other users.

So. An additional note; most security checks will advise activating the "PHP open_basedir Tweak" within WHM's security toolset. This is an automated tool for deploying the same item @imagina refers to - but in an automated fashion across all Apache virtualhosts. Like @imagina we have confirmed that disabling this protection will stop the SSL key error in Engintron from occurring. 

I guess it's up to individual hosts to decide how to approach this from a security vs. utility perspective. I don't think we saw this prior to cPanel v.66, so there is definitely a possibility that cPanel strengthened their security model or made adjustments as part of their new PHP-FPM work that caused this issue to surface.

In any event - we wished to retain the protections of open_basedir, so we created the following script to check if NGINX is currently running, and if not restart it, and if that fails (in our case usually due to SSL key mismatch) rebuild Engintron (mainline build) on the fly to clear the mismatch. As this error always occurs for us in the middle of the night during update runs an NGINX rebuild does not typically impact our clients. YMMV. Also note that we run CentOS 7, so if you run 6 be sure to modify the syntax for checking the service status, etc. accordingly.

https://pastebin.com/u8X8UEpM

The script (which we installed to /scripts/nginx_check.sh) is tapped every 5 minutes via CRON. Just enough time in our estimation to routinely verify NGINX hasn't died while at the same time giving Engintron enough breathing room to complete an install/rebuild on servers with a few hundred hosts deployed. 

`*/5 * * * * root /scripts/nginx_check.sh > /dev/null 2>&1`
